snuba/clickhouse/translator/rules.py

```diff
@@ -0,0 +1,85 @@
+from dataclasses import dataclass
+from typing import Optional
+
+from snuba.clickhouse.query import Expression
+from snuba.datasets.plans.translator.visitor import ExpressionMapper
+from snuba.query.dsl import array_element
+from snuba.query.expressions import (
+    Column,
+    ExpressionVisitor,
+    FunctionCall,
+    SubscriptableReference,
+)
+
+
+@dataclass(frozen=True)
+class ColumnMapper(ExpressionMapper[Column, Expression]):
+    """
+    Maps a column with a name and a table into a column with a different name and table.
+
+    The alias is not transformed.
+    """
+
+    from_col_name: str
+    from_table_name: Optional[str]
+    to_col_name: str
+    to_table_name: Optional[str]
+
+    def attemptMap(
+        self, expression: Column, children_translator: ExpressionVisitor[Expression],
+    ) -> Optional[Expression]:
+        if (
+            expression.column_name == self.from_col_name
+            and expression.table_name == self.from_table_name
+        ):
+            return Expression(
+                Column(
+                    alias=expression.alias,
+                    table_name=self.to_table_name,
+                    column_name=self.to_col_name,
+                )
+            )
+        else:
+            return None
+
+
+@dataclass(frozen=True)
+class TagMapper(ExpressionMapper[SubscriptableReference, Expression]):
+    """
+    Basic implementation of a tag mapper that transforms a subscriptable
```

Comment:
Does "subscriptable" relate only to nested columns? Or should ColumnMapper get passed a SubscriptableReference like TagMapper?

Reply:
SubscriptableReference is only loosely connected to the concept of nested column. It means an expression that can be accessed as a subscriptable (like [] in pyrthon). Today this is only for some very specific nested columns (tags) but it may be a simple not nested array in the future.
So yes, today it only relates to some specific nested columns, but it may become more generic.
Remember an expression in the AST is a vehicle to access a database concept (like a column), it does not represent the database concept itself.
So while a nested column is a column in the database, a SubscriptableReference in the ast is not a reference to a Column (thus it is not a Column), it is just an expression to access a value in a mapping column. Does this make sense ?

ColumnMapper cannot be passed a SubscriptableReference, it takes a different type in. This is intentional, the two classes are siblings.

Reply:
Yes, thanks